### PR TITLE
Update native vector provider to use unsigned int7 values only

### PR DESCRIPTION
**Pull request by ChrisHegarty**
This commit updates the native vector provider to reflect that Lucene's scalar quantization is unsigned int7, with a range of values from 0 to 127 inclusive. Stride has been pushed down into native, to allow other platforms to more easily select there own stride length.

Previously the implementation supports signed int8. We might want the more general signed int8 implementation in the future, but for now unsigned int7 is sufficient, and allows to provide more efficient implementations on x64.

### CODE REVIEW

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Review by ldematte — APPROVED** [review]
A couple of small comments but looks good to me! (if CI pass of course)

File: libs/vec/native/src/vec/c/vec.c [hunk 1]
```diff
@@ -46,15 +46,7 @@ EXPORT int vec_caps() {
...
-EXPORT int dot8s_stride() {
-    return DOT8_STRIDE_BYTES_LEN;
-}
-
-EXPORT int sqr8s_stride() {
-    return SQR8S_STRIDE_BYTES_LEN;
-}
-
-EXPORT int32_t dot8s(int8_t* a, int8_t* b, size_t dims) {
+int32_t dot7u_inner(int8_t* a, int8_t* b, size_t dims) {
```
[the remark this answers is not in the document]
Comment (ChrisHegarty) [reply libs/vec/native/src/vec/c/vec.c]: yes, that's better

File: libs/vec/native/src/vec/c/vec.c [hunk 2]
```diff
@@ -88,13 +80,26 @@ EXPORT int32_t dot8s(int8_t* a, int8_t* b, size_t dims) {
...
+        i += dims & ~(DOT7U_STRIDE_BYTES_LEN - 1);
+        res = dot7u_inner(a, b, i);
+    }
+    for (; i < dims; i++) {
+        res += a[i] * b[i];
+    }
+    return res;
+}
+
+EXPORT int32_t sqr7u_inner(int8_t *a, int8_t *b, size_t dims) {
```
[the remark this answers is not in the document]
Comment (ChrisHegarty) [reply libs/vec/native/src/vec/c/vec.c]: mistake. fixed.

**Comment by ChrisHegarty** [comment]
Performance is largely undisturbed by this change:

```
Benchmark                               (dims)   Mode  Cnt   Score   Error   Units
VectorScorerBenchmark.dotProductLucene    1024  thrpt    5   6.456 ± 0.029  ops/us
VectorScorerBenchmark.dotProductNative    1024  thrpt    5  28.232 ± 0.647  ops/us
VectorScorerBenchmark.dotProductScalar    1024  thrpt    5   3.022 ± 0.009  ops/us
```

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Pull request merged**